### PR TITLE
APIMF-2775: add custom resource loader documentation

### PR DESCRIPTION
**Pull request by arielmirra**
APIMF-2775: add custom resource loader documentation

### CODE REVIEW

File: docs/custom/resource-loaders.mdx [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+- URI of main raml api: `file:///Users/user/api.raml`
+- Reference to json file: `type: !include ./folder/../external file.json`
+
+This URI is normalized by the 3 steps described above:
+
+- Encode string
+    - the URI is transformed to `./folder/../external%20file.json`. Notice the encoding of the blank space between words.
+- Normalization
+    - the URI is transformed to `external%20file.json` by removing the relative path and folder structure
```
Comment (IsaacEldridge) [inline docs/custom/resource-loaders.mdx:87]: The relative path and folder structure are removed, the URI is transformed to `external%20file.json`.

File: docs/custom/resource-loaders.mdx [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+- Resolve relativeness
+    - as the reference was relative, this step appends the URI of the main file to the base of the reference URI: `file:///Users/user/external%20file.json`
```
Comment (IsaacEldridge) [inline docs/custom/resource-loaders.mdx:89]: - The reference was relative, so the URI of the of the main file is appended to the base of the reference URI: `file:///Users/user/external%20file.json`.

File: docs/custom/resource-loaders.mdx [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+Finally, resource loaders will receive the following the normalized URI to fetch the json file: `file:///Users/user/external%20file.json`.
```
Comment (IsaacEldridge) [inline docs/custom/resource-loaders.mdx:91]: Finally, resource loaders will use the following normalized URI to fetch the JSON file: `file:///Users/user/external%20file.json`.

File: docs/custom/resource-loaders.mdx [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+## Code example
```
Comment (IsaacEldridge) [inline docs/custom/resource-loaders.mdx:93]: ## Example code for a custom resource loader

File: docs/custom/resource-loaders.mdx [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+The following code is an example of a custom Resource Loader that is able to obtain resources using a custom protocol:
```
Comment (IsaacEldridge) [inline docs/custom/resource-loaders.mdx:94]: The following example code is for a custom resource loader that obtains resources using a custom protocol:

**Review by IsaacEldridge — CHANGES_REQUESTED** [review]

File: docs/custom/resource-loaders.mdx [hunk 6]
```diff
@@ -0,0 +1,121 @@
+---
+id: resource_loaders
+title: Custom resource loaders
+---
+
+AMF provides different resource loaders depending on the platform it's being used:
+
+- JVM
+  - FileResourceLoader
```
[the remark this answers is not in the document]
Comment (arielmirra) [reply docs/custom/resource-loaders.mdx:9]: was hesitating to do it, but they're literally things inside the AMF code, so yes

File: docs/custom/resource-loaders.mdx [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+#### Accepts
+`accepts(resource: string): boolean`
+The accept method receives the resource URI and determines if it's a resource it can fetch.
+AMF asks every resource loader in the environment through this method if it can fetch the current resource,
+and throws and `UnsupportedUrlScheme` exception if no resource loader can.
+
+#### Fetch
+`fetch(resource: string): ClientFuture[Content]`
+The fetch method encapsulates the actual loading of the resource, which depends on your implementation, but always returns a future (or promise) of a Content object.
```
[the remark this answers is not in the document]
Comment (arielmirra) [reply docs/custom/resource-loaders.mdx:44]: That's almost correct! Promise/Observable/Future/etc. are all implementations in different languages of asynchronous calls. I've reformatted the text to generalize that. thanks!

File: docs/custom/resource-loaders.mdx [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+The `Content` object is located at `amf.client.remote.Content` and has 3 things:
+
+- The resource fetched as a stream
+- The URL from which the resource was fetched
+- The [media type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types) of the resource
+
+
+## How do you configure your custom resource loader?
+All parsers receive an optional Environment parameter in their constructor.
+Within this Environment you can directly add your custom resource loaders, or even delete the ones provided by AMF:
```
[the remark this answers is not in the document]
Comment (arielmirra) [reply docs/custom/resource-loaders.mdx:55]: True, but the process of deleting the ones provided by AMF is not a straightforward one, I'll delete that part and add it if a refactor is made.

File: docs/custom/resource-loaders.mdx [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+```java title="CustomResourceLoader.java"
+Environment env = DefaultEnvironment.apply().add(new CustomResourceLoader());
+Raml10Parser parser = new Raml10Parser(env);
+```
+
+## URI normalization when calling resource loaders
+Before fetching, every URI goes through processing called **normalization** to ensure consistency and standardization.
```
[the remark this answers is not in the document]
Comment (arielmirra) [reply docs/custom/resource-loaders.mdx:63]: Much better, thanks

**Review by looseale — DISMISSED** [review]

**Review by IsaacEldridge — APPROVED** [review]

**Pull request merged**